### PR TITLE
rgw_file: implement reliable has-children check (unlink dir)

### PR DESCRIPTION
**Pull request by mattbenjamin**
Bug report and discussion provided by
Gui Hecheng <guihecheng@cmss.chinamobile.com> in nfs-ganesha upstream
github.  Briefly, while a reliable check is potentially costly,
it is necessary.

Fixes: http://tracker.ceph.com/issues/19270

Signed-off-by: Matt Benjamin <mbenjamin@redhat.com>

### CODE REVIEW

**Comment by mattbenjamin** [comment]
@cbodley I think this is ready to go--I found and fixed the...leaked mutex :)

File: src/rgw/rgw_file.h [hunk 1]
```diff
@@ -1373,6 +1374,89 @@ class RGWListBucketsRequest : public RGWLibRequest,
...
+
+  RGWRMdirCheck (CephContext* _cct, RGWUserInfo *_user,
+		 const RGWFileHandle* _rgw_fh)
+    : RGWLibRequest(_cct, _user), rgw_fh(_rgw_fh), valid(false),
+      has_children(false) {
+    default_max = 2;
+    op = this;
+  }
+
+  virtual bool only_bucket() { return false; }
```
Comment (cbodley) [inline src/rgw/rgw_file.h]: `bool only_bucket() override` please. same with overrides below

File: src/rgw/rgw_file.h [hunk 2]
```diff
@@ -1373,6 +1374,89 @@ class RGWListBucketsRequest : public RGWLibRequest,
...
+    return 0;
+  }
+
+  virtual int get_params() {
+    max = default_max;
+    return 0;
+  }
+
+  virtual void send_response() {
+    valid = true;
```
Comment (cbodley) [inline src/rgw/rgw_file.h]: what is this `valid` flag for? are there cases where the call to `rgwlib.get_fe()->execute_req(&req);` will return success without calling through here?

File: src/rgw/rgw_file.h (hunk 1 above)
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: right, thanks

File: src/rgw/rgw_file.h (hunk 2 above)
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: Potentially, yes.  This is send_response().

**Pull request merged**